modules/metastore/src/Controller/MetastoreAccessManager.php

```diff
@@ -0,0 +1,250 @@
+   *
+   * Right now this class only supports a single bundle per factory,
+   * although in theory the factory supports a mapping of schema IDs
+   * to bundles. In the future, this could be extended to support
+   * multiple bundles if needed.
+   */
+  protected string $bundle;
+
+  /**
+   * Constructor.
+   */
+  public function __construct(
+    EntityTypeManagerInterface $entityTypeManager,
+    MetastoreEntityItemFactoryInterface $itemFactory,
+  ) {
+    $this->itemFactory = $itemFactory;
+    $this->entityType = $itemFactory::getEntityType();
+    $bundles = $itemFactory::getBundles();
+    $this->bundle = reset($bundles);
+    $this->entityTypeManager = $entityTypeManager;
+    $this->accessControlHandler = $entityTypeManager->getAccessControlHandler($this->entityType);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('entity_type.manager'),
+      $container->get('dkan.metastore.metastore_item_factory'),
+    );
+  }
+
+  /**
+   * Check if user can create an item from a schema.
+   *
+   * @param string $schema_id
+   *   The schema ID.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The user account.
+   *
+   * @return \Drupal\Core\Access\AccessResult
+   *   An access result object indicating whether the user can create an item.
+   */
+  public function canCreate(string $schema_id, AccountInterface $account): AccessResult {
+    return $this->accessControlHandler->createAccess($this->bundle, $account, [], TRUE);
+  }
+
+  /**
+   * Check if user can update an item from a schema.
+   *
+   * @param string $schema_id
+   *   The schema ID.
+   * @param string $identifier
+   *   The item ID.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The user account.
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The HTTP request object.
+   *
+   * @return \Drupal\Core\Access\AccessResult
+   *   An access result object indicating whether the user can update the item.
+   */
+  public function canUpdate(
+    string $schema_id,
+    string $identifier,
+    AccountInterface $account,
+    Request $request,
+  ): AccessResult {
+    $method = $request->getMethod();
+    // Check if the user has permission to update items of this schema.
+    try {
+      $entity = $this->getEntity($schema_id, $identifier);
+      return $this->accessControlHandler->access($entity, "update", $account, TRUE);
+    }
+    catch (MissingObjectException | \InvalidArgumentException $e) {
+      // If this is a PUT, we should check if user has CREATE permission.
+      if ($method === 'PUT' && !$this->canCreate($schema_id, $account)->isAllowed()) {
+        return AccessResult::forbidden();
+      }
+      // If the the item does not exist, assume "allowed" and let the controller
+      // handle the 404 response.
+      return AccessResult::allowed();
```

Comment:
It seems weird to see this in code, tho I know we discussed it. It seems like missing object would be a different outcome than invalid argument, but I guess not.

It looks like we have this pattern covered in `DatasetItemTest` so 👍